src/vsr/grid_repair_queue.zig

```diff
@@ -0,0 +1,548 @@
+        index_block_data: BlockPtrConst,
+    ) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_blocks.count() ==
+            queue.enqueued_single_blocks + queue.enqueued_table_blocks);
+        assert(table.table_blocks_total == null);
+        assert(table.table_blocks_written == 0);
+        assert(table.content_blocks_received.count() == 0);
+
+        const index_schema = schema.TableIndex.from(index_block_data);
+        const index_block_header = schema.header_from_block(index_block_data);
+        assert(index_block_header.op == table.index_address);
+        assert(index_block_header.checksum == table.index_checksum);
+        assert(schema.BlockType.from(index_block_header.operation) == .index);
+
+        const content_blocks_total = index_schema.content_blocks_used(index_block_data);
+        table.table_blocks_total = 1 + content_blocks_total;
+
+        var content_block_index: u32 = 0;
+        while (content_block_index < content_blocks_total) : (content_block_index += 1) {
+            const block_id = index_schema.content_block(index_block_data, content_block_index);
+
+            const enqueue = queue.enqueue_faulty_block(
+                block_id.block_address,
+                block_id.block_checksum,
+                .{ .table_content = .{ .table = table, .index = content_block_index } },
+            );
+
+            if (enqueue == .replace) {
+                if (enqueue.replace.state == .writing) {
+                    table.content_blocks_received.set(content_block_index);
+                }
+            } else {
+                assert(enqueue == .insert);
+            }
+        }
+    }
+
+    fn release_fault(queue: *GridRepairQueue, fault_index: usize) void {
+        assert(!queue.canceling);
+        assert(queue.faulty_blocks_repair_index < queue.faulty_blocks.count());
+
+        switch (queue.faulty_blocks.entries.items(.value)[fault_index].progress) {
+            .block => queue.enqueued_single_blocks -= 1,
+            .table_index => queue.enqueued_table_blocks -= 1,
+            .table_content => queue.enqueued_table_blocks -= 1,
+        }
+
+        queue.faulty_blocks.swapRemoveAt(fault_index);
+
+        if (queue.faulty_blocks_repair_index == queue.faulty_blocks.count()) {
+            queue.faulty_blocks_repair_index = 0;
+        } else {
+            if (queue.faulty_blocks_repair_index > fault_index) {
+                queue.faulty_blocks_repair_index -= 1;
+            }
+        }
+    }
+
+    pub fn cancel(queue: *GridRepairQueue) void {
+        assert(!queue.canceling);
+        assert(queue.checkpointing == null);
+
+        queue.faulty_blocks.clearRetainingCapacity();
+        queue.faulty_blocks_repair_index = 0;
+        queue.enqueued_single_blocks = 0;
+        queue.enqueued_table_blocks = 0;
+
+        queue.canceling = true;
+        while (queue.faulty_tables.pop()) |table| {
+            (table.callback)(table, .canceled);
```

Comment:
Hm, invoking callbacks here looks suspicious. In `Replica.sync_request_table_callback`, it's not clear why we won't enter

```
            if (self.sync_tables) |_| {
                self.sync_request_tables();
            }
```

and then re-enter GridRepairQueue. Let's explicitly match over `RepairTableResult` in `sync_request_table_callback` and assert that we don't do anything funny in the canceled case.

Reply:
I think on cancellation we _will_ enter `sync_request_tables()`, and that is good -- we are queuing tables that belong to the checkpoint (since they are in our manifest levels), so they are safe to repair concurrent to the checkpoint.

Reply:
Oh wait nevermind, I was thinking of the checkpoint case, didn't realize this was cancellation!